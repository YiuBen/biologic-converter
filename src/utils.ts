import { PartialFileList } from './Types';

/**
 * Interface for [[`groupFiles`]]
 */
export interface GroupedFiles {
  /** group id used to group the files under same path */
  gid: string;
  /** file name as taken from [[`File`]] or compatible class.*/
  name: string;
  /** file extension if it has an extension */
  extension?: string;
  /** there is no other way atm, this will store the files */
  // eslint-disable-next-line
  [file: string]: any;
}

/**
 * Options for [[`groupFiles`]]
 * If neither `useBasename` nor `useExtension` are set it uses the Filename.
 */
export interface GroupFilesOptions {
  /** A path like `./dir1/dir2/base.ext` will use `./dir1/dir2/base` as id.
   */
  idWithBasename?: boolean;
  /** The `./dir1/dir2/base.ext` will be stored under `base` key */
  useBasename?: boolean;
  /** The `./dir1/dir2/base.ext` will be stored under `ext` key.
   * If no extension is found in file, it will use the basename. */
  useExtension?: boolean;
  /**use basename and extension for the name.*/
  useFilename?: boolean;
  /** both extension and  basename are lowercased, it doesnt touch the `id` field. */
  lowerCaseFileKey?: boolean;
}

/**
 * Group files under same directory path
 * i.e `[Dir1, Dir2, Dir3...]`  where each Dir object ooks like this
 * ```
 * {
 *   id,
 *   extension,
 *   name,
 *   fileA,
 *   fileB,...
 *  }
 *  ```
 * See `[[GroupFilesOptions]]` for all options.
 * Files _must be_ in the same directory to be grouped.
 * @param fileList
 * @return - array that _may_ contain `{mpr, mps and mpt}` as [[`File`]] Objects.
 */
export function groupFiles(
  fileList: PartialFileList | FileList,
  options: GroupFilesOptions = {
    idWithBasename: true,
    useExtension: true,
  },
): GroupedFiles[] {
  let { idWithBasename, useExtension, useBasename, lowerCaseFileKey } = options;

  let results: Record<string, Partial<GroupedFiles>> = {};

  for (const file of fileList) {
    /* set up name variables */
    let filename = file.name;
    if (lowerCaseFileKey) filename = filename.toLowerCase();
    const basename = filename.replace(/\.[^.]+$/, '');
    const extension = filename.replace(/.*\./, '');
    let gid;

    /* set the unique grouping id */
    if (idWithBasename) {
      gid = file.webkitRelativePath.replace(/\.[^/]+$/, '');
    } else {
      //just dir path, like for WDF parser.
      gid = file.webkitRelativePath.replace(/\.[^.]+$/, '');
    }

    /* create the partial group object inside results */
    if (!results[gid]) {
      Object.assign(results, { [gid]: { gid } });
    }

    /* place file under a particular group `results[gid]` */
    if (useExtension) {
      //if no extension, extension = basename anyways
      Object.assign(results[gid], { [extension]: file });
    } else if (useBasename) {
      //by basename
      Object.assign(results[gid], { [basename]: file });
    } else {
      // by fullname
      Object.assign(results[gid], { [filename]: file });
    }
  }
  return Object.keys(results).map((gid) => results[gid]) as GroupedFiles[];
}
