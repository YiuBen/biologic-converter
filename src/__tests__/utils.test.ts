//temporary till groupfiles is released
import { join } from 'path';

import { fileListFromPath } from 'filelist-from';

import { groupFiles } from '../utils';

describe('file grouping at a complex directory', () => {
  it('all data', () => {
    const files = fileListFromPath(join(__dirname, 'data/'));
    const result = groupFiles(files, {
      idWithBasename: true,
      useExtension: true,
    });
    expect(result).toHaveLength(3);
    //expect(result).toMatchSnapshot();
  });

  it('jdb11-4', () => {
    const files = fileListFromPath(join(__dirname, 'data/jdb11-4'));
    const result = groupFiles(files, {
      idWithBasename: false,
      useExtension: false,
      useBasename: true,
      lowerCaseFileKey: true,
    });
    expect(result).toHaveLength(1);
  });

  it('test file', () => {
    const files = fileListFromPath(join(__dirname, 'data/test'));
    const result = groupFiles(files, {
      idWithBasename: false,
      useExtension: false,
      useBasename: false,
      useFilename: true,
    });
    console.log(result);
    expect(Object.keys(result[0])).toHaveLength(4);
  });
});
