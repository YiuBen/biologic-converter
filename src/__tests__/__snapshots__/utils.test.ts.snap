// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`file grouping at a complex directory all data 1`] = `
Array [
  Object {
    "gid": "/home/sm/clones/biologic-converter/src/__tests__/data/jdb11-1/jdb11-1",
    "mpr": Object {
      "arrayBuffer": [Function],
      "lastModified": 1647170961688,
      "name": "jdb11-1.mpr",
      "size": 2465718,
      "text": [Function],
      "webkitRelativePath": "/home/sm/clones/biologic-converter/src/__tests__/data/jdb11-1/jdb11-1.mpr",
    },
    "mps": Object {
      "arrayBuffer": [Function],
      "lastModified": 1647170961688,
      "name": "jdb11-1.mps",
      "size": 4330,
      "text": [Function],
      "webkitRelativePath": "/home/sm/clones/biologic-converter/src/__tests__/data/jdb11-1/jdb11-1.mps",
    },
  },
  Object {
    "gid": "/home/sm/clones/biologic-converter/src/__tests__/data/jdb11-4/jdb11-4",
    "mpr": Object {
      "arrayBuffer": [Function],
      "lastModified": 1647170961704,
      "name": "jdb11-4.mpr",
      "size": 2041303,
      "text": [Function],
      "webkitRelativePath": "/home/sm/clones/biologic-converter/src/__tests__/data/jdb11-4/jdb11-4.mpr",
    },
    "mps": Object {
      "arrayBuffer": [Function],
      "lastModified": 1647170961704,
      "name": "jdb11-4.mps",
      "size": 9648,
      "text": [Function],
      "webkitRelativePath": "/home/sm/clones/biologic-converter/src/__tests__/data/jdb11-4/jdb11-4.mps",
    },
  },
  Object {
    "gid": "/home/sm/clones/biologic-converter/src/__tests__/data/test/test",
    "mpr": Object {
      "arrayBuffer": [Function],
      "lastModified": 1647170961704,
      "name": "test.mpr",
      "size": 379643,
      "text": [Function],
      "webkitRelativePath": "/home/sm/clones/biologic-converter/src/__tests__/data/test/test.mpr",
    },
    "mps": Object {
      "arrayBuffer": [Function],
      "lastModified": 1647515352914,
      "name": "test.mps",
      "size": 1911,
      "text": [Function],
      "webkitRelativePath": "/home/sm/clones/biologic-converter/src/__tests__/data/test/test.mps",
    },
    "mpt": Object {
      "arrayBuffer": [Function],
      "lastModified": 1647170961712,
      "name": "test.mpt",
      "size": 1695957,
      "text": [Function],
      "webkitRelativePath": "/home/sm/clones/biologic-converter/src/__tests__/data/test/test.mpt",
    },
  },
]
`;
